lib/AST/TypeJoinMeet.cpp

```diff
@@ -17,100 +17,197 @@
 #include "swift/AST/ASTContext.h"
 #include "swift/AST/Decl.h"
 #include "swift/AST/Type.h"
+#include "swift/AST/TypeVisitor.h"
 #include "swift/AST/Types.h"
 #include "llvm/ADT/SmallPtrSet.h"
 using namespace swift;
 
-Type Type::join(Type type1, Type type2) {
-  assert(!type1->hasTypeVariable() && !type2->hasTypeVariable() &&
-         "Cannot compute join of types involving type variables");
-
-  assert(type1->getWithoutSpecifierType()->isEqual(type1) &&
-         "Expected simple type!");
-  assert(type2->getWithoutSpecifierType()->isEqual(type2) &&
-         "Expected simple type!");
-
-  // FIXME: This algorithm is woefully incomplete, and is only currently used
-  // for optimizing away extra exploratory work in the constraint solver. It
-  // should eventually encompass all of the subtyping rules of the language.
-
-  // If the types are equivalent, the join is obvious.
-  if (type1->isEqual(type2))
-    return type1;
-
-  // If both are class metatypes, compute the join of the instance type and
-  // wrap the result in a metatype.
-  if (auto *metatype1 = type1->getAs<MetatypeType>()) {
-    if (auto *metatype2 = type2->getAs<MetatypeType>()) {
-      auto instance1 = metatype1->getInstanceType();
-      auto instance2 = metatype2->getInstanceType();
-      if (instance1->mayHaveSuperclass() &&
-          instance2->mayHaveSuperclass()) {
-        auto result = Type::join(instance1, instance2);
-        if (!result)
-          return result;
-        return MetatypeType::get(result);
-      }
-    }
+// FIXME: This is currently woefully incomplete, and is only currently
+// used for optimizing away extra exploratory work in the constraint
+// solver. It should eventually encompass all of the subtyping rules
+// of the language.
+struct TypeJoin : TypeVisitor<TypeJoin, Type> {
```

Comment:
I definitely appreciate the use of `TypeVisitor` here to break out the cases. It's moderately annoying that we have to deal with asymmetry at each level, but even [`TypeMatcher`](https://github.com/apple/swift/blob/master/include/swift/AST/TypeMatcher.h) would I guess be worse at that.

Reply:
Yeah I looked at `TypeMatcher` and concluded that it would probably be no better and possibly worse. Not to mention it is based on `CanTypeVisitor` which explicitly disallows unchecked types.

I was going to ask you: Do you happen to know whether there was a good reason for that?.

Reply:
I think `CanTypeVisitor` might have come from a part of the compiler that didn't handle unchecked types. I don't see a reason for that restriction now.